internal/watch/file_watcher.go

```diff
@@ -60,15 +62,31 @@ func (f *OSFileWatcher) watch() error {
 	}
 	f.watcher = watcher
 
+	realFile, err := filepath.EvalSymlinks(f.file)
+	if err != nil {
+		return err
+	}
+
 	dir, file := path.Split(f.file)
 	go func(file string) {
 		for {
 			select {
 			case event := <-watcher.Events:
-				if (event.Op&fsnotify.Write == fsnotify.Write ||
-					event.Op&fsnotify.Create == fsnotify.Create) &&
-					strings.HasSuffix(event.Name, file) {
-					f.onEvent()
+				if event.Op&fsnotify.Create == fsnotify.Create ||
+					event.Op&fsnotify.Write == fsnotify.Write {
+					if finfo, err := os.Lstat(event.Name); err != nil {
+						log.Printf("can not lstat file: %v\n", err)
+					} else if finfo.Mode()&os.ModeSymlink != 0 {
+						if currentRealFile, err := filepath.EvalSymlinks(f.file); err == nil &&
+							currentRealFile != realFile {
+							f.onEvent()
```

Comment:
Do you not have to condition this on strings.HasSuffix(event.Name, file)? Why?

Reply:
the FileWatcher watches the entire directory (that's not new behaviour). Using the example in the PR description, the watcher will watch $dir/ca.crt, which is a symlink to $dir/..data/ca.crt, which is a symlink to $dir/..<dated_dir>/ca.crt. The `CREATE` event in $dir is not for a file called `ca.crt` (which would have the correct suffix), but for a **directory** called `..data` which symlinks to the dated directory. However, the watcher should detect that the top-level symlink (`$dir/ca.crt`) now points to a different file, and trigger an event.
tl;dr when any file is created in the directory being watched, this FileWatcher should confirm if the symlink it was told to monitor has been updated somewhere within the directory (maybe deeper than 1 level).

Reply:
More details: [this comment](https://github.com/spf13/viper/pull/415#issuecomment-344050279) shows the internal order of events.